### PR TITLE
feat: ignore `collate` in `select` and `create table`

### LINKED ISSUE #13171: Support (ignore) `collate "c"`
### Is your feature request related to a problem? Please describe.

This syntax is used in Dreamio - Postgres connector. 

Basically, `encode "UTF-8" collate "c"` means comparing string by their UTF-8 encoded bytes, without any collation.

> When lc_collate is "C" (or "POSIX"), the comparison is done internally by PostgreSQL. In that case, it compares the byte representations of the strings using memcmp.

This is the same as Rust string's comparison, which is also the behavior of RisingWave. So we can safely ignore it.

> Strings are compared [lexicographically](https://doc.rust-lang.org/std/cmp/trait.Ord.html#lexicographical-comparison) by their byte values.

References:

- https://dba.stackexchange.com/questions/240930/postgresql-difference-between-collations-c-and-c-utf-8
- https://doc.rust-lang.org/std/primitive.str.html

### Describe the solution you'd like

Support it in parser but ignore the `collate "c"` in binder. For other `collate`, may throw an error.

We may limit the scope to `collate "c"` in `select` queries only, ignore the `create table` with collate.

### Describe alternatives you've considered

No idea.

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by jetjinser**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

We don't support real collations, but we ignore it when the collation is "C" or "POSIX" as it is RisingWave's current default behavior.
The purpose of this PR is to make dremio work properly. resolve #13171.

When writing e2e test, I discovered a parser problem: `select varchar 't' collate "C"` cannot be parsed normally.
In order to make this PR easier, the parser problem will be solved in other PRs in the future.

## Checklist

- [ ] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

</details>

### CODE REVIEW

File: src/frontend/src/binder/expr/mod.rs [hunk 1]
```diff
@@ -561,6 +562,18 @@ impl Binder {
         let lhs = self.bind_expr_inner(expr)?;
         lhs.cast_explicit(data_type).map_err(Into::into)
     }
+
+    pub fn bind_collate(&mut self, expr: Expr, collation: ObjectName) -> Result<ExprImpl> {
+        if !["C", "POSIX"].contains(&collation.real_value().to_uppercase().as_str()) {
```
Comment (xiangjinwu) [inline src/frontend/src/binder/expr/mod.rs:567]: PostgreSQL only accepts uppercase:
```
test=# select 'abc' collate "C";
 ?column? 
----------
 abc
(1 row)

test=# select 'abc' collate "c";
ERROR:  collation "c" for encoding "UTF8" does not exist
LINE 1: select 'abc' collate "c";
                     ^
test=# select 'abc' collate C; -- when outside double quotes, `real_value` calls `to_lowercase`
ERROR:  collation "c" for encoding "UTF8" does not exist
LINE 1: select 'abc' collate C;
                     ^
```

File: e2e_test/batch/basic/collation.slt.part [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+----
+t a
+
+query B
+select 'a' < ('b' collate "C");
+----
+t
+
+query B
+select ('a' < 'b') collate "C";
```
Comment (xiangjinwu) [inline e2e_test/batch/basic/collation.slt.part:57]: This is invalid in PostgreSQL for similar reasons as the `create_table` handler comment: only Varchar is collatable

File: src/frontend/src/binder/expr/mod.rs (hunk 1 above)
Comment (jetjinser) [reply src/frontend/src/binder/expr/mod.rs:567]: Oh, I made a mistake, you are right

File: src/frontend/src/binder/expr/mod.rs [hunk 3]
```diff
@@ -561,6 +562,18 @@ impl Binder {
...
+    pub fn bind_collate(&mut self, expr: Expr, collation: ObjectName) -> Result<ExprImpl> {
+        if !["C", "POSIX"].contains(&collation.real_value().as_str()) {
+            return Err(ErrorCode::NotImplemented(
+                "Collate collation other than `C` or `POSIX` is not implemented".into(),
+                None.into(),
+            )
+            .into());
+        }
+
+        self.bind_expr_inner(expr)
```
Comment (xiangjinwu) [inline src/frontend/src/binder/expr/mod.rs:575]: If we bind the inner expr first, we can `if bound_inner.return_type() != DataType::Varchar`. The `create_table` handler has access to `data_type` as well. Not required, but seems easy to achieve.

Btw found an interesting case in PostgreSQL:
```
test=# select '1' collate "C" > 2;
 ?column? 
----------
 f
(1 row)
```
It considers `'1'` as collatable initially, then converts it into an interger for `>`. Our implementation behaves the same way.

**Review by xiangjinwu — APPROVED** [review]

**Pull request merged**